Documentation/MyFirstContribution

```diff
@@ -0,0 +1,584 @@
+
+For the purposes of this doc, we will base all our work on `master`. Before
+running the command below, ensure that you are on `master` first so your
+branch diverges at the right point.
+
+----
+git checkout -b psuh
+----
+
+We'll make a number of commits here in order to demonstrate how to send many
+patches up for review simultaneously.
+
+== Code It Up!
+
+NOTE: A reference implementation can be found at
+https://github.com/nasamuffin/git/tree/codelab.
+
+=== Adding a new command
+
+Lots of the main useful commands are written as builtins. So it is informative
+to implement `git psuh` as a builtin.
+
+Create a new file in `builtin/` called `psuh.c`.
+
+The entry point of your new command needs to match a certain signature:
+
+----
+int cmd_psuh(int argc, const char **argv, const char *prefix)
+----
+
+We'll also need to add the extern declaration of psuh; open up `builtin.h`,
+find the declaration for cmd_push, and add a new line for psuh:
+
+----
+extern int cmd_psuh(int argc, const char **argv, const char *prefix);
+----
+
+Be sure to `#include "builtin.h"` in your `psuh.c`.
+
+Go ahead and add some throwaway printf to that method. This is a decent
+starting point as we can now add build rules and register the command.
+
+Let's try to build it.  Open Makefile, find where `builtin/push.o` is added
+to BUILTIN_OBJS, and add `builtin/psuh.o` in the same way. Once you've done so,
+move to the root directory and build simply with `make -j16`. Optionally, add
+the DEVELOPER=1 variable to turn on some additional warnings:
+
+----
+echo DEVELOPER=1 > config.mak
+make -j16
+----
+
+Great, now your new command builds happily on its own. But nobody invokes it.
+Let's change that.
+
+The list of commands lives in `git.c`. We can register a new command by adding
+a cmd_struct to the commands[] array. struct cmd_struct takes a string with the
+command name, a function pointer to the command implementation, and a setup
+option flag. For now, let's keep cheating off of push. Find the line where
+cmd_push is registered, copy it, and modify it for cmd_psuh. Leave the options
+the same as for push.
+
+TODO: what do the options do? How to fix them?
+
+Go ahead and build again. You should see a clean build, so let's kick the tires
+and see if it works. There's a binary you can use to test with in
+`./bin-wrappers`.
+
+----
+./bin-wrappers/git psuh
+----
+
+Check it out! You've got a command! Nice work!  Let's commit this.
+
+----
+git add Makefile builtin.h builtin/psuh.c git.c
+git commit -sm "builtin: add hooks for git-psuh"
+----
+
+Go ahead and inspect your new commit with `git show`. "builtin:" indicates you
+have modified mainly the builtin/ area. The subject line gives readers an idea
+of what you've changed. The signed-off line (-s) indicates that you agree to
+the Developer's Certificate of Origin 1.1 (see the SubmittingPatches [[dco]]
+header). If you wish to add some context to your change, go ahead with
+`git commit --amend`.
```

Comment:
As I mentioned in comments on your commits, the commit messages really need to bear information in order to be accepted by the Git project. Maybe that should be clarified here, together with a good example?

Reply:
Reworked the commit messages.